gix-utils/src/btoi.rs

```diff
@@ -0,0 +1,316 @@
+///
+/// assert!(to_signed::<u8>(b"123456789").is_err()); // overflow
+/// assert!(to_signed::<u8>(b"-1").is_err()); // underflow
+///
+/// assert!(to_signed::<i32>(b" 42").is_err()); // leading space
+/// ```
+pub fn to_signed<I: MinNumTraits>(bytes: &[u8]) -> Result<I, ParseIntegerError> {
+    to_signed_with_radix(bytes, 10)
+}
+
+/// Converts a byte slice in a given base to an integer.
+///
+/// Like [`to_unsigned_with_radix`], but numbers may optionally start with a sign
+/// (`-` or `+`).
+///
+/// # Errors
+///
+/// Returns [`ParseIntegerError`] for any of the following conditions:
+///
+/// * `bytes` has no digits
+/// * not all characters of `bytes` are `0-9`, `a-z`, `A-Z`, exluding an
+///   optional leading sign
+/// * not all characters refer to digits in the given `radix`, exluding an
+///   optional leading sign
+/// * the number overflows or underflows `I`
+///
+/// # Panics
+///
+/// Panics if `radix` is not in the range `2..=36` (or in the pathological
+/// case that there is no representation of `radix` in `I`).
+///
+/// # Examples
+///
+/// ```
+/// # use gix_utils::btoi::to_signed_with_radix;
+/// assert_eq!(Ok(10), to_signed_with_radix(b"a", 16));
+/// assert_eq!(Ok(10), to_signed_with_radix(b"+a", 16));
+/// assert_eq!(Ok(-42), to_signed_with_radix(b"-101010", 2));
+/// ```
+pub fn to_signed_with_radix<I: MinNumTraits>(bytes: &[u8], radix: u32) -> Result<I, ParseIntegerError> {
+    assert!(
+        (2..=36).contains(&radix),
+        "radix must lie in the range 2..=36, found {radix}"
+    );
+
+    let base = I::from_u32(radix).expect("radix can be represented as integer");
+
+    if bytes.is_empty() {
+        return Err(ParseIntegerError { kind: ErrorKind::Empty });
+    }
+
+    let digits = match bytes[0] {
+        b'+' => return to_unsigned_with_radix(&bytes[1..], radix),
+        b'-' => &bytes[1..],
+        _ => return to_unsigned_with_radix(bytes, radix),
+    };
+
+    if digits.is_empty() {
+        return Err(ParseIntegerError { kind: ErrorKind::Empty });
+    }
+
+    let mut result = I::zero();
+
+    for &digit in digits {
+        let x = match char::from(digit).to_digit(radix).and_then(I::from_u32) {
+            Some(x) => x,
+            None => {
+                return Err(ParseIntegerError {
+                    kind: ErrorKind::InvalidDigit,
+                })
+            }
+        };
+        result = match result.checked_mul(base) {
+            Some(result) => result,
+            None => {
+                return Err(ParseIntegerError {
+                    kind: ErrorKind::Underflow,
+                })
+            }
+        };
+        result = match result.checked_sub(x) {
+            Some(result) => result,
+            None => {
+                return Err(ParseIntegerError {
+                    kind: ErrorKind::Underflow,
+                })
+            }
+        };
+    }
+
+    Ok(result)
+}
+
+/// minimal subset of traits used by btoi_radix and btou_radix
```

Comment:
```suggestion
/// minimal subset of traits used by [`to_signed_with_radix `] and [`to_signed_with_radix`]
```
or however it is that doclinks (or just generally improve this doc?)

Reply:
That's a great catch, thanks for the suggestion.
I didn't apply the fix because it uses `to_signed_with_radix` twice, but applied the intended fix locally and force-pushed.